app/scripts/components/academy/academy-course.components.jsx

```diff
@@ -0,0 +1,337 @@
+							onChange={() => {this.markAsRead(props.children);}}
+							/>
+						<label htmlFor={`${this.courseSlug}-${props.children}`}><span></span>{props.children}</label>
+					</div>
+			);
+		}
+		else {
+			return(
+				this.createElement(`h${props.level}`, this.getCoreProps(props), props.children)
+			);
+		}
+	}
+
+	createCourseProgress() {
+		const academyProgress = this.state.academyProgress || {};
+
+		this.state.headers && this.state.headers.map((header) => {
+			const parts = [];
+
+			parts.push({
+				name: header.content,
+				completed: false,
+			});
+			this.client.dispatchAction(
+				'/create-course-progress',
+				{
+					slug: tutorial.slug,
+					name: tutorial.title,
+					parts,
+				}
+			);
+			academyProgress[tutorial.slug] = {
+				parts,
+				rewarded: false,
+				name: tutorial.title,
+			};
+		});
+		this.setState({academyProgress});
+	}
+
+	markAsRead(part) {
+		this.client.dispatchAction('/mark-part-as-read', {course: this.courseSlug, part: part[0]});
+	}
+
+	isPartRead(part) {
+		let coursePart = {};
+
+		if (this.state.academyProgress
+		&& this.state.academyProgress[this.courseSlug]) {
+			coursePart = this.state.academyProgress[this.courseSlug].parts.find((elem) => {
+				return elem.name === part[0];
+			});
+		}
+		return coursePart.completed;
+	}
+
+	areAllPartsRead() {
+		if (!this.state.academyProgress || !this.state.academyProgress[this.courseSlug]) {
+			return false;
+		}
+
+		const partsDone = this.state.academyProgress[this.courseSlug].parts.filter((part) => {
+			return part.completed === true;
+		}).length || 0;
+
+		return (
+			partsDone === this.state.academyProgress[this.courseSlug].parts.length
+		);
+	}
+
+	getNextCourse() {
+		return this.tutorials.content[this.tutorials.content.findIndex((tutorial) => {
+			return tutorial.slug === this.courseSlug;
+		}) + 1];
+	}
+
+	render() {
+		const renderers = {
+			Heading: this.headerRenderer,
+		};
+		const course = this.tutorials.content.find((tutorial) => {
+			return tutorial.slug === this.courseSlug;
+		});
+		const parts = course.content.split("## ");
+
+		parts.map((part, index) => {
+			if (index !== 0) {
+				parts[index] = `## ${part}`;
+			}
+		});
+
+		const basics = course.basics.length > 0 ? (
+			<div>
+				<h3>Basics</h3>
+				{course.basics.map((basic) => {
+					return (
+						<Link className="academy-sidebar-menu-item" to={`/academy/course/${basic.slug}`} > {basic.title} </Link>
+					);
+				})}
+			</div>
+		) : null;
+
+		const sidebar = (
+			<div>
+				{basics}
+				<h3>Parts</h3>
+				{this.state.headers ? this.state.headers.map((header) => {
+					return (
+						<span
+							className={`academy-sidebar-menu-item ${header.active ? 'is-active' : ''}`}
+							onClick={() => {header.elem.scrollIntoView();}}>
+							{header.content}
+						</span>
+					);
+				}) : false}
+			</div>
+		);
+
+		return(
+			<div key={this.courseName} className="academy-base academy-course" ref="academyCoursePage">
```

Comment:
Yup, that's what they do https://facebook.github.io/react/docs/refs-and-the-dom.html#adding-a-ref-to-a-class-component